### PR TITLE
Non partitioned container support for V3 SDK

### PR DESCRIPTION
**Pull request by srinathnarayanan**
- Added partitionKey.Empty to be used while reading, deleting or querying a non partitioned collection using API v 2018-12-31
- Added support for adding the right partition key for create, replace and upsert of documents into non partitioned containers using API v 2018-12-31
- Bumped API  version to 2018-12-31
- Added sample for above scenarios

### CODE REVIEW

File: azure/cosmos/cosmos_client_connection.py [hunk 1]
```diff
@@ -2882,20 +2883,23 @@ def _ExtractPartitionKey(self, partitionKeyDefinition, document):
...
+        is_system_key = (partitionKeyDefinition['systemKey']
+                         if 'systemKey' in partitionKeyDefinition else False)
 
         # Navigates the document to retrieve the partitionKey specified in the paths
-        return self._RetrievePartitionKey(partition_key_parts, document)
+        return self._RetrievePartitionKey(partition_key_parts, document, is_system_key)
 
     # Navigates the document to retrieve the partitionKey specified in the partition key parts
-    def _RetrievePartitionKey(self, partition_key_parts, document):
+    def _RetrievePartitionKey(self, partition_key_parts, document, is_system_key):
```
Comment (christopheranderson) [inline azure/cosmos/cosmos_client_connection.py]:
```suggestion
    def _retrieve_partition_key(self, partition_key_parts, document, is_system_key):
```

File: azure/cosmos/cosmos_client_connection.py [hunk 2]
```diff
@@ -2952,3 +2956,11 @@ def _get_database_link(database_or_id):
         else:
             database_id = cast("Dict[str, str]", database_or_id)["id"]
         return "dbs/{}".format(database_id)
+
+    @staticmethod
+    def _ReturnUndefinedOrEmptyPartitionKey(is_system_key):
```
Comment (christopheranderson) [inline azure/cosmos/cosmos_client_connection.py]:
```suggestion
    def _return_undefined_or_empty_partition_key(is_system_key):
```

File: samples/NonPartitionedCollectionOperations/Program.py [hunk 3]
```diff
@@ -0,0 +1,316 @@
+import azure.cosmos.cosmos_client as cosmos_client
```
Comment (christopheranderson) [inline samples/NonPartitionedCollectionOperations/Program.py]: missing header

File: azure/cosmos/partition_key.py [hunk 4]
```diff
@@ -20,6 +20,21 @@
 #SOFTWARE.
 
 
+class NonePartitionKeyValue(object):
+    """Represents none value for partitionKey when it's missing in a containers.
+    """
+
+
+class _Empty(object):
```
Comment (christopheranderson) [inline azure/cosmos/partition_key.py]: make sure these types (_Empty, _Undefined) don't leak from azure.cosmos

File: test/partition_key_tests.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    host = test_config._test_config.host
+    masterKey = test_config._test_config.masterKey
+    connectionPolicy = test_config._test_config.connectionPolicy
+    client = cosmos_client.CosmosClient(host, {'masterKey': masterKey}, "Session", connectionPolicy)
+    created_db = test_config._test_config.create_database_if_not_exist(client)
+    created_collection = test_config._test_config.create_multi_partition_collection_with_custom_pk_if_not_exist(client)
+
+    @classmethod
+    def tearDownClass(cls):
```
Comment (christopheranderson) [inline test/partition_key_tests.py]:
```suggestion
    def tear_down_class(cls):
```

File: test/partition_key_tests.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        cls.created_db.delete_container(container=cls.created_collection_id)
+
+    @classmethod
+    def setUpClass(cls):
```
Comment (christopheranderson) [inline test/partition_key_tests.py]:
```suggestion
    def set_up_class(cls):
```

File: test/test_config.py [hunk 7]
```diff
@@ -38,7 +38,7 @@ class _test_config(object):
 
     #[SuppressMessage("Microsoft.Security", "CS002:SecretInNextLine", Justification="Cosmos DB Emulator Key")]
     masterKey = os.getenv('ACCOUNT_KEY', 'C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==')
-    host = os.getenv('ACCOUNT_HOST', 'https://localhost:443')
+    host = os.getenv('ACCOUNT_HOST', 'https://localhost:443')
```
Comment (christopheranderson) [inline test/test_config.py]: any reason for this extra space?

File: test/partition_key_tests.py (hunk 5 above)
Comment (srinathnarayanan) [reply test/partition_key_tests.py]: overrides baseclass method from unittest.Testcase, should not be renamed

File: test/partition_key_tests.py (hunk 6 above)
Comment (srinathnarayanan) [reply test/partition_key_tests.py]: overrides baseclass method from unittest.Testcase, should not be renamed

File: test/test_config.py (hunk 7 above)
Comment (srinathnarayanan) [reply test/test_config.py]: had to send a extra commit to trigger a build from the PR, reverting it

**Pull request merged**